src/Simplex/Chat.hs

```diff
@@ -231,48 +231,120 @@ processAgentMessage user@User {userId, profile} agentConnId agentMessage = do
+              case chatMsgEvent of
+                XGrpAcpt memId
+                  | memId == memberId m -> do
+                    withStore $ \st -> updateGroupMemberStatus st userId (groupMemberId m) GSMemAccepted
+                    withAgent $ \a -> allowConnection a agentConnId confId ""
+                  | otherwise -> pure () -- TODO error not matching member ID
+                _ -> pure () -- TODO show/log error, other events in SMP confirmation
+            _ -> do
+              -- TODO this member should send `x.grp.mem.info` in SMP confirmation
+              -- user can compare role with the previously sent by host and only accept if they match
+              -- TODO unsolved problem is how to manage role updates that can happen concurrently with introductions
+              pure ()
         CON -> do
-          Group {membership, members} <- withStore $ \st -> getGroup st user gName
-          -- TODO because the contact is not created instantly, if the member is not created from contact,
-          -- it should still have a unique local display name.
-          -- If it is created from contact it can still be duplicated on the member (and match the contact)
-          case invitedBy m of
-            IBUser -> do
-              -- sender was invited by the current user
-              withStore $ \st -> updateGroupMemberStatus st userId (groupMemberId m) GSMemConnected
-              sendGroupMessage members $ XGrpMemNew (memberId m) (memberRole m) (memberProfile m)
+          Group {members} <- withStore $ \st -> getGroup st user gName
+          withStore $ \st -> updateGroupMemberStatus st userId (groupMemberId m) GSMemConnected
+          case memberCategory m of
+            GCInviteeMember -> do
+              -- TODO createIntroductions should create pending intros for all "known" members
+              -- but only members that are already connected should be sent in XGrpMemNew, so possibly
+              -- createIntroductions should return [GroupMember] and here it would be filtered and sorted
               showConnectedGroupMember gName $ displayName (memberProfile m :: Profile)
-              forM_ (filter (\m' -> memberStatus m' >= GSMemConnected) . map fst $ members) $ \m' ->
-                sendDirectMessage agentConnId $ XGrpMemIntro (memberId m') (memberRole m') (memberProfile m')
-            _ -> do
-              if Just (invitedBy membership) == (IBContact <$> memberContactId m)
-                then do
-                  -- sender invited the current user
-                  withStore $ \st -> updateGroupMemberStatus st userId (groupMemberId m) GSMemConnected
-                  showUserConnectedToGroup gName
-                  pure ()
-                else do
-                  showConnectedGroupMember gName $ displayName (memberProfile m :: Profile)
+              mIds <- withStore $ \st -> createIntroductions st m
+              sendGroupMessage members . (`XGrpMemNew` mIds) $ memberInfo m
```

Comment:
TODO update status to "announced"

Reply:
I decided to use announced only for members that are not connected yet (that is neither host nor invitee)